gdplogd/log_graph.go

```diff
@@ -0,0 +1,272 @@
+package gdplogd
+
+import (
+	"database/sql"
+	"fmt"
+
+	_ "github.com/mattn/go-sqlite3"
+)
+
+// LogGraph descriobe the graph of a DataCapsule on a log server
+type LogGraph struct {
+	// Address of the DataCapsule
+	GraphAddr HashAddr
+
+	db *sql.DB
+
+	// All log entries in the database as of last refresh
+	logEntries []LogEntryMetadata
+
+	forwardEdges  HashAddrMultiMap
+	backwardEdges map[HashAddr]HashAddr
+	logicalEnds   []HashAddr
+	logicalBegins []HashAddr
+
+	// The actual hash pointer map, which follows:
+	// A (oldest) <- B <- C (newest)
+	HashPtrMap map[HashAddr]HashAddr
+}
+
+func (logGraph *LogGraph) GetActualPtrMap() map[HashAddr]HashAddr {
+	return logGraph.backwardEdges
+}
+
+func (logGraph *LogGraph) GetLogicalPtrMap() HashAddrMultiMap {
+	return logGraph.forwardEdges
+}
+
+func (logGraph *LogGraph) GetLogicalEnds() []HashAddr {
+	return logGraph.logicalEnds
+}
+
+func (logGraph *LogGraph) GetLogicalBegins() []HashAddr {
+	return logGraph.logicalBegins
+}
+
+// Return LogGraph and calculate its logical represntation
+func InitLogGraph(graphAddr HashAddr, db *sql.DB) (LogGraph, error) {
+	var logGraph LogGraph
+	logGraph.GraphAddr = graphAddr
+	logGraph.db = db
+
+	err := logGraph.RefreshLogGraph()
+	if err != nil {
+		return logGraph, err
+	}
+
+	return logGraph, nil
+}
+
+// Update log properties from the logs database.
+// Specifically updates logEntries, forwardEdges, backwardEdges
+func (logGraph *LogGraph) RefreshLogGraph() error {
```

Comment:
I think adding this function to `LogGraphWrapper` or perhaps a more piecemeal version is good for controlling the recency of this data. Thoughts?